File: backend/services/tcgdex_languages.py
"""Supported TCGdex language helpers.

Keep language validation, ordering, and composite ID suffix parsing in one place so
new TCGdex languages do not require scattered en/de checks.
"""

from __future__ import annotations

import re
from collections.abc import Iterable
from typing import Any

DEFAULT_TCGDEX_SYNC_LANGUAGES: tuple[str, ...] = ("en", "de")
ENGLISH_FALLBACK_LANGUAGE = "en"
SUPPORTED_TCGDEX_LANGUAGES: tuple[str, ...] = (
    "en",
    "fr",
    "es",
    "es-mx",
    "it",
    "pt",
    "pt-br",
    "pt-pt",
    "de",
    "nl",
    "pl",
    "ru",
    "ja",
    "ko",
    "zh-tw",
    "id",
    "th",
    "zh-cn",
)

_LANGUAGE_ALIASES = {
    "zh": "zh-cn",
    "zh-hans": "zh-cn",
    "zh_hans": "zh-cn",
    "zh-cn": "zh-cn",
    "zh_cn": "zh-cn",
    "zh-hant": "zh-tw",
    "zh_hant": "zh-tw",
    "zh-tw": "zh-tw",
    "zh_tw": "zh-tw",
    "jp": "ja",
    "kr": "ko",
    "br": "pt-br",
}

_LANGUAGE_SUFFIX_RE = re.compile(
    r"_(" + "|".join(re.escape(lang) for lang in sorted(SUPPORTED_TCGDEX_LANGUAGES, key=len, reverse=True)) + r")$"
)


def normalize_tcgdex_language(value: Any) -> str:
    """Normalize one language code for TCGdex lookups."""
    code = str(value or "").strip().lower().replace("_", "-")
    return _LANGUAGE_ALIASES.get(code, code)


def is_supported_tcgdex_language(value: Any) -> bool:
    return normalize_tcgdex_language(value) in SUPPORTED_TCGDEX_LANGUAGES


def iter_tcgdex_language_parts(value: Any) -> list[str]:
    """Return normalized language parts from CSV/list input, preserving repeats."""
    if value is None:
        return []
    if isinstance(value, str):
        parts = re.split(r"[,\s]+", value)
    elif isinstance(value, Iterable):
        parts = []
        for item in value:
            if isinstance(item, str):
                parts.extend(re.split(r"[,\s]+", item))
            else:
                parts.append(str(item))
    else:
        parts = [str(value)]
    return [normalize_tcgdex_language(part) for part in parts if str(part or "").strip()]


def normalize_tcgdex_sync_languages(value: Any, *, default_on_empty: bool = True) -> str:
    """Normalize configured sync languages to a stable CSV string.

    Unknown values are ignored so Docker/env typos cannot break startup. If no
    valid values remain and default_on_empty is true, the default en,de pair is
    returned. The special value "all" expands to every known TCGdex language.
    """
    raw_parts = iter_tcgdex_language_parts(value)
    if "all" in raw_parts:
        selected = list(SUPPORTED_TCGDEX_LANGUAGES)
    else:
        selected = []
        for lang in SUPPORTED_TCGDEX_LANGUAGES:
            if lang in raw_parts and lang not in selected:
                selected.append(lang)

    if not selected and default_on_empty:
        selected = list(DEFAULT_TCGDEX_SYNC_LANGUAGES)

    return ",".join(selected)


def validate_tcgdex_sync_languages(value: Any) -> str:
    """Normalize user-provided sync languages or raise ValueError if none are valid."""
    normalized = normalize_tcgdex_sync_languages(value, default_on_empty=False)
    if not normalized:
        allowed = ", ".join(SUPPORTED_TCGDEX_LANGUAGES)
        raise ValueError(f"tcgdex_sync_languages must include at least one supported language: {allowed}")
    return normalized


def strip_lang_suffix(value: str | None) -> tuple[str, str]:
    """Return (base_id, lang) for composite IDs like sv1-1_zh-tw.

    Legacy IDs without a supported suffix default to English for backward
    compatibility.
    """
    identifier = str(value or "")
    match = _LANGUAGE_SUFFIX_RE.search(identifier)
    if match:
        lang = match.group(1)
        return identifier[: match.start()], lang
    return identifier, ENGLISH_FALLBACK_LANGUAGE


def has_lang_suffix(value: str | None) -> bool:
    return bool(_LANGUAGE_SUFFIX_RE.search(str(value or "")))


def with_lang_suffix(base_id: str, lang: str) -> str:
    return f"{base_id}_{normalize_tcgdex_language(lang)}"


def english_fallback_languages(target_lang: str | None) -> list[str]:
    """Preferred data fallback order for missing card/set data.

    English is the global fallback source. For English rows there is no stronger
    fallback source, so return an empty list instead of guessing from another
    language.
    """
    lang = normalize_tcgdex_language(target_lang)
    if lang and lang != ENGLISH_FALLBACK_LANGUAGE:
        return [ENGLISH_FALLBACK_LANGUAGE]
    return []


def supported_tcgdex_language_payload() -> list[dict[str, str]]:
    names = {
        "en": "English",
        "fr": "French",
        "es": "Spanish",
        "es-mx": "Spanish Mexico",
        "it": "Italian",
        "pt": "Portuguese",
        "pt-br": "Portuguese Brazil",
        "pt-pt": "Portuguese Portugal",
        "de": "German",
        "nl": "Dutch",
        "pl": "Polish",
        "ru": "Russian",
        "ja": "Japanese",
        "ko": "Korean",
        "zh-tw": "Chinese Traditional",
        "id": "Indonesian",
        "th": "Thai",
        "zh-cn": "Chinese Simplified",
    }
    return [{"code": code, "name": names[code]} for code in SUPPORTED_TCGDEX_LANGUAGES]

File: frontend/src/utils/tcgdexLanguages.js
export const DEFAULT_TCGDEX_SYNC_LANGUAGES = ['en', 'de']

export const TCGDEX_LANGUAGES = [
  { code: 'en', name: 'English', short: 'EN', flag: '🇬🇧' },
  { code: 'fr', name: 'French', short: 'FR', flag: '🇫🇷' },
  { code: 'es', name: 'Spanish', short: 'ES', flag: '🇪🇸' },
  { code: 'es-mx', name: 'Spanish Mexico', short: 'ES-MX', flag: '🇲🇽' },
  { code: 'it', name: 'Italian', short: 'IT', flag: '🇮🇹' },
  { code: 'pt', name: 'Portuguese', short: 'PT', flag: '🇵🇹' },
  { code: 'pt-br', name: 'Portuguese Brazil', short: 'PT-BR', flag: '🇧🇷' },
  { code: 'pt-pt', name: 'Portuguese Portugal', short: 'PT-PT', flag: '🇵🇹' },
  { code: 'de', name: 'German', short: 'DE', flag: '🇩🇪' },
  { code: 'nl', name: 'Dutch', short: 'NL', flag: '🇳🇱' },
  { code: 'pl', name: 'Polish', short: 'PL', flag: '🇵🇱' },
  { code: 'ru', name: 'Russian', short: 'RU', flag: '🇷🇺' },
  { code: 'ja', name: 'Japanese', short: 'JA', flag: '🇯🇵' },
  { code: 'ko', name: 'Korean', short: 'KO', flag: '🇰🇷' },
  { code: 'zh-tw', name: 'Chinese Traditional', short: 'ZH-TW', flag: '🇹🇼' },
  { code: 'id', name: 'Indonesian', short: 'ID', flag: '🇮🇩' },
  { code: 'th', name: 'Thai', short: 'TH', flag: '🇹🇭' },
  { code: 'zh-cn', name: 'Chinese Simplified', short: 'ZH-CN', flag: '🇨🇳' },
]

const LANGUAGE_BY_CODE = new Map(TCGDEX_LANGUAGES.map((language) => [language.code, language]))

const LANGUAGE_BADGE_CLASSES = [
  'bg-blue-500/20 text-blue-300 border border-blue-500/30',
  'bg-pink-500/20 text-pink-300 border border-pink-500/30',
  'bg-orange-500/20 text-orange-300 border border-orange-500/30',
  'bg-amber-500/20 text-amber-300 border border-amber-500/30',
  'bg-emerald-500/20 text-emerald-300 border border-emerald-500/30',
  'bg-teal-500/20 text-teal-300 border border-teal-500/30',
  'bg-lime-500/20 text-lime-300 border border-lime-500/30',
  'bg-cyan-500/20 text-cyan-300 border border-cyan-500/30',
  'bg-yellow/20 text-yellow border border-yellow/30',
  'bg-sky-500/20 text-sky-300 border border-sky-500/30',
  'bg-violet-500/20 text-violet-300 border border-violet-500/30',
  'bg-red-500/20 text-red-300 border border-red-500/30',
  'bg-rose-500/20 text-rose-300 border border-rose-500/30',
  'bg-purple-500/20 text-purple-300 border border-purple-500/30',
  'bg-fuchsia-500/20 text-fuchsia-300 border border-fuchsia-500/30',
  'bg-green-500/20 text-green-300 border border-green-500/30',
  'bg-indigo-500/20 text-indigo-300 border border-indigo-500/30',
  'bg-slate-500/20 text-slate-300 border border-slate-500/30',
]

const LANGUAGE_BADGE_CLASS_BY_CODE = new Map(
  TCGDEX_LANGUAGES.map((language, index) => [language.code, LANGUAGE_BADGE_CLASSES[index] || LANGUAGE_BADGE_CLASSES[0]])
)

export function normalizeTcgdexLanguage(value, fallback = 'en') {
  const code = String(value || '').trim().toLowerCase().replace(/_/g, '-')
  const aliases = {
    zh: 'zh-cn',
    'zh-hans': 'zh-cn',
    'zh-hant': 'zh-tw',
    jp: 'ja',
    kr: 'ko',
    br: 'pt-br',
  }
  const normalized = aliases[code] || code
  return LANGUAGE_BY_CODE.has(normalized) ? normalized : fallback
}

export function getTcgdexLanguage(code) {
  return LANGUAGE_BY_CODE.get(normalizeTcgdexLanguage(code)) || LANGUAGE_BY_CODE.get('en')
}

export function tcgdexLanguageLabel(code, { full = false } = {}) {
  const language = getTcgdexLanguage(code)
  return full ? `${language.flag} ${language.name}` : `${language.flag} ${language.short}`
}

export function tcgdexLanguageBadgeClass(code) {
  return LANGUAGE_BADGE_CLASS_BY_CODE.get(normalizeTcgdexLanguage(code)) || LANGUAGE_BADGE_CLASSES[0]
}

export function normalizeTcgdexLanguageCsv(value) {
  const rawParts = String(value || '')
    .split(/[\s,]+/)
    .map((part) => normalizeTcgdexLanguage(part, ''))
    .filter(Boolean)
  const selected = TCGDEX_LANGUAGES
    .map((language) => language.code)
    .filter((code) => rawParts.includes(code))
  return selected.length ? selected.join(',') : DEFAULT_TCGDEX_SYNC_LANGUAGES.join(',')
}
